### ISSUE TITLE
[Bug]: Unable to create exactly-once Flink pipeline with stream source and file sink

### ISSUE THREAD

**Issue by psolomin** [apache/beam#26041]
### What happened?

My pipeline is consuming from streaming source (AWS Kinesis) and writes to parquet files once per minute.

When restarting it from a Flink snapshot, I was unable to get exactly-once delivery of the records into the final files.

1. When using standard file-naming policy, I get files like `out-2023-03-29T22-45-00.000Z-2023-03-29T22-46-00.000Z-00000-of-00004.parquet`. I've noticed that, after a restart, some of the window files are replaced, which causes records to be lost.
2. I've created a JobId var which is re-generated every time the app starts, which gives file names like `out-<JobId>-2023-03-29T22-45-00.000Z-2023-03-29T22-46-00.000Z-00000-of-00004.parquet` - this allows me to avoid having replaced files and lost records, but introduces duplicates. The file names are made via https://github.com/psolomin/beam-playground/blob/e0f1834d6ee76a796d6d21836e1d04140e6d9ca2/kinesis-io-with-enhanced-fan-out/src/main/java/com/psolomin/consumer/NoColonFileNaming.java#L16

(2) effectively makes the pipeline to be at-least-once.

Also, when the app fails and Flink restarts it, I don't see neither duplicates, nor lost records, and exactly-once semantic actually works as expected.

I've tried different setting of the window:

```
Window.<KinesisRecord>into(FixedWindows.of(Duration.standardSeconds(60)))
```

```
Window.<KinesisRecord>into(FixedWindows.of(Duration.standardSeconds(60)))
    .withAllowedLateness(Duration.ZERO)   <- changing this to greater numbers had no effect
    .discardingFiredPanes()
       .triggering(Repeatedly.forever(AfterProcessingTime.pastFirstElementInPane()
       .plusDelayOf(Duration.standardSeconds(60))))
```

Is it expected behaviour? Is it documented anywhere? Is there a configuration combo of savepoint config and window config which preserves exaclty-once when the app is stopped and started from a savepoint?

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [X] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

**Comment by psolomin** [comment]
> Without making any additional actions

Correct. I tried to reproduce data loss again - and it does happen if I omit my `runId` in file names - files are replaced with new files when the app starts from a savepoint. Here's my code:

https://github.com/psolomin/beam-playground/tree/file-sink-app-id-trick/kinesis-io-with-enhanced-fan-out#vanilla-flink

I still think that it's my window misconfiguration, but windows are configured like that in all examples, including GCP one: https://cloud.google.com/pubsub/docs/stream-messages-dataflow#stream_messages_from_to

> If so, this should be P1

I would say, it should be. If this is expected behaviour, I think, windowing documentation should explain how windows behave when app is stopped and then re-started from serialised state + include window configuration examples which allow safe re-start from serialised state.

**Comment by kennknowles** [comment]
I wonder if this has to do with `RequiresStableInput`. I do know that there are places where it is required for randomly generated UUIDs to be checkpointed and replayed deterministically. I don't know enough about Flink or the FlinkRunner to know why restarting from checkpoint would be different from an app restart. I also don't know enough to know if this is more general to the IO connector.

**Comment by gabrywu** [comment]
any updates here? I encounter the same issue. And I think it should be a P1 issue.
```java
2023-08-07 13:17:43,263 INFO  org.apache.beam.sdk.io.FileBasedSink                         [] - Will copy temporary file FileResult{tempFilename=hdfs://tmp/.iceberg/.temp-beam/204b6d82ff4a1e7e-6a72-4cae-ab63-5f48cb2053c7, shard=7, window=[2023-08-07T16:30:00.000Z..2023-08-07T16:40:00.000Z), paneInfo=PaneInfo{isFirst=true, isLast=true, timing=ON_TIME, index=0, onTimeIndex=0}} to final location hdfs://tmp/.iceberg/part-1691425800000-1691426400000-07-12.snappy.iceberg.parquet


2023-08-07 14:50:51,739 INFO  org.apache.beam.sdk.io.FileBasedSink                         [] - Will copy temporary file FileResult{tempFilename=hdfs://tmp/.iceberg/.temp-beam/118a0b0a366e6090-a180-4a58-9de9-586a80b9e97d, shard=7, window=[2023-08-07T16:30:00.000Z..2023-08-07T16:40:00.000Z), paneInfo=PaneInfo{isFirst=true, isLast=true, timing=ON_TIME, index=0, onTimeIndex=0}} to final location hdfs://tmp/.iceberg/part-1691425800000-1691426400000-07-12.snappy.iceberg.parquet
```
